csrc/device_lower/utils.cpp

```diff
@@ -942,6 +942,59 @@ bool isReductionInitExpr(const Expr* expr) {
   return true;
 }
 
+// Return true if it is sufficient to predicate the end of the loop
+// iteration. An aligned vectorized loop is one example where it is
+// guaranteed to be valid by the validation checks. More generally,
+// the divisible split set is used to find such loops. The divisible
+// split set contains splits used in view transformations as well as
+// those whose output domains are vectorized. View transformations
+// guarantee that any split involved is divisible, whereas
+// vectorization only guarantees that the overall root extent is
+// divisible by the split factor. Thus, if a loop IterDomain is
+// an output of a split included in the divisible view splits, we can
+// just predicate the end of the loop iteration. If a loop IterDomain
+// is an output of a divisible split due to vectorization, it is only
+// valid when the loop IterDomain is mapped with the vectorized inner
+// output IterDomain. If it is mapped with an outer IterDomain, since
+// the split input IterDomain may be an output IterDomain of a
+// non-divisible split, we still need to predicate each loop iteration
+// value.
+bool predicateAtEnd(ForLoop* loop) {
+  auto loop_id = loop->iter_domain();
+  auto split = dynamic_cast<Split*>(loop_id->definition());
+  if (split == nullptr) {
+    return false;
+  }
```

Comment:
I realize this is pre-existing code, but wouldn't it also be possible to have a divisible split followed by a merge? For example:
```
tv[ i0, 32 ]
tv->split(1, 8); // [i0, 4, 8]
tv->split(2, 4); // [i0, 4, 2, 4]
tv->merge(1, 3); // [i0, 16 , 2]
tv->axis(0)->parallelize(ParallelType::Unswitch);
```
Now the both of the dimensions in `tv` divide evenly into the logical dimension of `tv`, but only one of them is a direct split output, so as I understand it that means only the inner loop will have `predicateAtEnd` return `true`. Is this correct, and is this the desired behavior?

Reply:
I think you're right. This check won't detect such patterns. Not ideal, but this was sufficient when I was working on some normalization or reduction fusions. You're more than welcome to extend it 😁